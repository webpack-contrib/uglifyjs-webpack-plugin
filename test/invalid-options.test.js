import UglifyJsPlugin from '../src/index';
import {
  PluginEnvironment,
  cleanErrorStack,
  createCompiler,
  compile,
} from './helpers';

describe('when applied with invalid options', () => {
  it('matches snapshot', () => {
    const compiler = createCompiler();
    new UglifyJsPlugin({
      uglifyOptions: {
        output: {
          'invalid-option': true,
        },
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      const errors = stats.compilation.errors.map(cleanErrorStack);
      const warnings = stats.compilation.warnings.map(cleanErrorStack);

      expect(errors).toMatchSnapshot('errors');
      expect(warnings).toMatchSnapshot('warnings');

      for (const file in stats.compilation.assets) {
        if (Object.prototype.hasOwnProperty.call(stats.compilation.assets, file)) {
          expect(stats.compilation.assets[file].source()).toMatchSnapshot(file);
        }
      }
    });
  });

  it('throws validation errors', () => {
    /* eslint-disable no-new */
    expect(() => {
      new UglifyJsPlugin({ test: /foo/ });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ doesntExist: true });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ cache: true });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ cache: 'path/to/cache/directory' });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ cache: {} });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ additionalCacheKeys: false });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ additionalCacheKeys: 'additional-cache-key' });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ additionalCacheKeys: { path: 'path/to/output' } });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ parallel: true });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ parallel: 2 });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ parallel: '2' });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ parallel: {} });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ sourceMap: true });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ sourceMap: 'true' });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: null });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { ie8: false } });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { ie8: true } });
    }).not.toThrow('Validation Error');

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { ie8: 'false' } });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { emca: 5 } });
    }).not.toThrow();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { emca: 8 } });
    }).not.toThrow();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { ecma: 7.5 } });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { ecma: true } });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { ecma: '5' } });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { ecma: 3 } });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      new UglifyJsPlugin({ uglifyOptions: { ecma: 10 } });
    }).toThrowErrorMatchingSnapshot();
  });

  it('outputs uglify errors', () => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin({
      uglifyOptions: {
        output: {
          'invalid-option': true,
        },
      },
    });
    plugin.apply(compilerEnv);
    const [eventBinding] = pluginEnvironment.getEventBindings();

    const chunkPluginEnvironment = new PluginEnvironment();
    const compilation = chunkPluginEnvironment.getEnvironmentStub();
    compilation.assets = {
      'test.js': {
        source: () => 'var foo = 1;',
      },
    };
    compilation.errors = [];

    eventBinding.handler(compilation);
    const [compilationEventBinding] = chunkPluginEnvironment.getEventBindings();

    compilationEventBinding.handler([{
      files: ['test.js'],
    }], () => {
      expect(compilation.errors.length).toBe(1);
      expect(compilation.errors[0]).toBeInstanceOf(Error);
      expect(compilation.errors[0].message).toEqual(expect.stringContaining('from UglifyJs'));
    });
  });
});
