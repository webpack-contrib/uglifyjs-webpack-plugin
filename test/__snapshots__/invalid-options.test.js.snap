// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when applied with invalid options matches snapshot: errors 1`] = `
Array [
  "Error: main.0c220ec66316af2c1b24.js from UglifyJs
DefaultsError: \`invalid-option\` is not a supported option",
  "Error: manifest.d6857f782c13a99b5917.js from UglifyJs
DefaultsError: \`invalid-option\` is not a supported option",
]
`;

exports[`when applied with invalid options matches snapshot: main.0c220ec66316af2c1b24.js 1`] = `
"webpackJsonp([0],[
/* 0 */
/***/ (function(module, exports) {

// foo
/* @preserve*/
// bar
const a = 2 + 2;

module.exports = function Foo() {
  const b = 2 + 2;
  console.log(b + 1 + 2);
};


/***/ })
],[0]);"
`;

exports[`when applied with invalid options matches snapshot: manifest.d6857f782c13a99b5917.js 1`] = `
"/******/ (function(modules) { // webpackBootstrap
/******/ 	// install a JSONP callback for chunk loading
/******/ 	var parentJsonpFunction = window[\\"webpackJsonp\\"];
/******/ 	window[\\"webpackJsonp\\"] = function webpackJsonpCallback(chunkIds, moreModules, executeModules) {
/******/ 		// add \\"moreModules\\" to the modules object,
/******/ 		// then flag all \\"chunkIds\\" as loaded and fire callback
/******/ 		var moduleId, chunkId, i = 0, resolves = [], result;
/******/ 		for(;i < chunkIds.length; i++) {
/******/ 			chunkId = chunkIds[i];
/******/ 			if(installedChunks[chunkId]) {
/******/ 				resolves.push(installedChunks[chunkId][0]);
/******/ 			}
/******/ 			installedChunks[chunkId] = 0;
/******/ 		}
/******/ 		for(moduleId in moreModules) {
/******/ 			if(Object.prototype.hasOwnProperty.call(moreModules, moduleId)) {
/******/ 				modules[moduleId] = moreModules[moduleId];
/******/ 			}
/******/ 		}
/******/ 		if(parentJsonpFunction) parentJsonpFunction(chunkIds, moreModules, executeModules);
/******/ 		while(resolves.length) {
/******/ 			resolves.shift()();
/******/ 		}
/******/ 		if(executeModules) {
/******/ 			for(i=0; i < executeModules.length; i++) {
/******/ 				result = __webpack_require__(__webpack_require__.s = executeModules[i]);
/******/ 			}
/******/ 		}
/******/ 		return result;
/******/ 	};
/******/
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// objects to store loaded and loading chunks
/******/ 	var installedChunks = {
/******/ 		1: 0
/******/ 	};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, {
/******/ 				configurable: false,
/******/ 				enumerable: true,
/******/ 				get: getter
/******/ 			});
/******/ 		}
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = \\"\\";
/******/
/******/ 	// on error function for async loading
/******/ 	__webpack_require__.oe = function(err) { console.error(err); throw err; };
/******/ })
/************************************************************************/
/******/ ([]);"
`;

exports[`when applied with invalid options matches snapshot: warnings 1`] = `Array []`;

exports[`when applied with invalid options throws validation errors 1`] = `
"UglifyJs Plugin Invalid Options

options['doesntExist'] is an invalid additional property
"
`;

exports[`when applied with invalid options throws validation errors 2`] = `
"UglifyJs Plugin Invalid Options

options.cache should be boolean
options.cache should be string
options.cache should match exactly one schema in oneOf
"
`;

exports[`when applied with invalid options throws validation errors 3`] = `
"UglifyJs Plugin Invalid Options

options.additionalCacheKeys should be object
"
`;

exports[`when applied with invalid options throws validation errors 4`] = `
"UglifyJs Plugin Invalid Options

options.additionalCacheKeys should be object
"
`;

exports[`when applied with invalid options throws validation errors 5`] = `
"UglifyJs Plugin Invalid Options

options.parallel should be boolean
options.parallel should be integer
options.parallel should match exactly one schema in oneOf
"
`;

exports[`when applied with invalid options throws validation errors 6`] = `
"UglifyJs Plugin Invalid Options

options.parallel should be boolean
options.parallel should be integer
options.parallel should match exactly one schema in oneOf
"
`;

exports[`when applied with invalid options throws validation errors 7`] = `
"UglifyJs Plugin Invalid Options

options.sourceMap should be boolean
"
`;

exports[`when applied with invalid options throws validation errors 8`] = `
"UglifyJs Plugin Invalid Options

options.uglifyOptions should be object
"
`;

exports[`when applied with invalid options throws validation errors 9`] = `
"UglifyJs Plugin Invalid Options

options.uglifyOptions.ie8 should be boolean
"
`;

exports[`when applied with invalid options throws validation errors 10`] = `
"UglifyJs Plugin Invalid Options

options.uglifyOptions.ecma should be integer
"
`;

exports[`when applied with invalid options throws validation errors 11`] = `
"UglifyJs Plugin Invalid Options

options.uglifyOptions.ecma should be integer
"
`;

exports[`when applied with invalid options throws validation errors 12`] = `
"UglifyJs Plugin Invalid Options

options.uglifyOptions.ecma should be integer
"
`;

exports[`when applied with invalid options throws validation errors 13`] = `
"UglifyJs Plugin Invalid Options

options.uglifyOptions.ecma should be >= 5
"
`;

exports[`when applied with invalid options throws validation errors 14`] = `
"UglifyJs Plugin Invalid Options

options.uglifyOptions.ecma should be <= 8
"
`;
